### PR TITLE
Add fast field for associating arbitrary bytes to a document

### LINKED ISSUE #270: Create `&[u8]` fast fields
They can use the double array strategy used for multivalued int.
This requires to add another type in the schema.

We can use base64 for its JSON encoding.

### PR DESCRIPTION
**Pull request by jason-wolfe**
Closes #270

### CODE REVIEW

**Comment by jason-wolfe** [comment]
WIP because I haven't actually made non-test changes that depend on this to test it for real.

**Comment by jason-wolfe** [comment]
Also still missing merging support.

File: src/fastfield/bytes/reader.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+/// Reader for byte array fast fields
+///
+/// The reader is implemented as a `u64` fast field and a separate collection of bytes.
+///
+/// The `vals_reader` will access the concatenated list of all values for all documents.
+///
+/// The `idx_reader` associates, for each document, the index of its first value.
+///
+/// Reading the value for a document is done by reading the start index for it,
+/// and the start index for the next document, and keeping the bytes in between.
```
Comment (fulmicoton) [inline src/fastfield/bytes/reader.rs:16]: Awesome doc.

File: src/fastfield/bytes/writer.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    /// Shift to the next document and add all of the
+    /// matching field values present in the document.
+    pub fn add_document(&mut self, doc: &Document) {
+        self.next_doc();
+        for field_value in doc.field_values() {
+            if field_value.field() == self.field {
+                if let &Value::Bytes(ref bytes) = field_value.value() {
+                    self.vals.extend_from_slice(bytes);
+                } else {
+                    panic!("MultiValued Bytes contained non-Bytes Value!: {:?}", field_value);
```
Comment (fulmicoton) [inline src/fastfield/bytes/writer.rs]: maybe add the field id in the error msg.

File: src/indexer/segment_writer.rs [hunk 3]
```diff
@@ -202,6 +202,9 @@ impl<'a> SegmentWriter<'a> {
                         }
                     }
                 }
+                FieldType::Bytes => {
+                    // Do nothing. Bytes only supports fast fields.
```
Comment (fulmicoton) [inline src/indexer/segment_writer.rs:206]: cool! THanks for not using `_`

File: src/schema/field_type.rs [hunk 4]
```diff
@@ -120,7 +139,7 @@ impl FieldType {
                         Err(ValueParsingError::OverflowError(msg))
                     }
                 }
-                FieldType::Str(_) | FieldType::HierarchicalFacet => {
+                FieldType::Str(_) | FieldType::HierarchicalFacet | FieldType::Bytes => {
```
Comment (fulmicoton) [inline src/schema/field_type.rs]: I think you want bytes as another case, and panic or return an error.

File: src/schema/field_type.rs [hunk 5]
```diff
@@ -135,3 +154,17 @@ impl FieldType {
...
+
+#[cfg(test)]
+mod tests {
+    use super::FieldType;
+    use schema::Value;
+
+    #[test]
+    fn test_bytes_value_from_json() {
+        let result = FieldType::Bytes
+            .value_from_json(&json!("dGhpcyBpcyBhIHRlc3Q="))
```
Comment (fulmicoton) [inline src/schema/field_type.rs:167]: cool. maybe test the errors as well.

File: src/schema/field_entry.rs [hunk 6]
```diff
@@ -168,8 +181,16 @@ impl<'de> Deserialize<'de> for FieldEntry {
                                 return Err(de::Error::duplicate_field("type"));
                             }
                             ty = Some(map.next_value()?);
-                            if ty == Some("hierarchical_facet") {
-                                field_type = Some(FieldType::HierarchicalFacet);
+                            match ty {
```
Comment (fulmicoton) [inline src/schema/field_entry.rs]: something is weird in this code (not yours). We wrap it in `Some` and match it right away?

Comment (fulmicoton) [reply src/schema/field_entry.rs]: Can you have a look at the bigger picture and see if there is something wrong here?

File: src/postings/postings_writer.rs [hunk 7]
```diff
@@ -36,7 +36,7 @@ fn posting_from_field_entry<'a>(
                 }
             })
             .unwrap_or_else(|| SpecializedPostingsWriter::<NothingRecorder>::new_boxed(heap)),
-        FieldType::U64(_) | FieldType::I64(_) | FieldType::HierarchicalFacet => {
+        FieldType::U64(_) | FieldType::I64(_) | FieldType::HierarchicalFacet | FieldType::Bytes => {
```
Comment (fulmicoton) [inline src/postings/postings_writer.rs]: `FieldType::Bytes` cannot be indexed at all. So you probably want a specific case.
`NothingRecorder` is a bit misleading here. It deos not "not index" the field. It indexes it with no extra informaiton (like docfreq or position)

File: src/schema/field_entry.rs (hunk 6 above)
Comment (jason-wolfe) [reply src/schema/field_entry.rs]: Yeah, will do.

File: src/postings/postings_writer.rs (hunk 7 above)
Comment (jason-wolfe) [reply src/postings/postings_writer.rs]: I'm not sure what the right solution is here.

One option is to return on `Option<Box<PostingsWriter + 'a>>` and never create a postings writer for the field. I don't know enough to know if this will cause downstream blowups.

Another option is to return a `ReallyDoesNothingRecorder` (obviously not the real name :smiley:).

File: src/schema/field_type.rs [hunk 8]
```diff
@@ -135,3 +154,32 @@ impl FieldType {
...
+        match result {
+            Err(ValueParsingError::TypeError(_)) => {}
+            _ => panic!("Expected parse failure for wrong type")
+        }
+
+        let result = FieldType::Bytes
+            .value_from_json(&json!("-"));
+        match result {
+            Err(ValueParsingError::InvalidBase64(_)) => {}
+            _ => panic!("Expected parse failure for invalid base64")
```
Comment (fulmicoton) [inline src/schema/field_type.rs:182]: awesome.

File: src/postings/postings_writer.rs (hunk 7 above)
Comment (fulmicoton) [reply src/postings/postings_writer.rs]: I think you are right. The whole thing will stay ugly as long as it does not under go a rather large refactoring.

File: src/schema/field_type.rs (hunk 5 above)
Comment (fulmicoton) [reply src/schema/field_type.rs:167]: thx

File: src/postings/postings_writer.rs (hunk 7 above)
Comment (fulmicoton) [reply src/postings/postings_writer.rs]: Can you just add the following comment :
// `FieldType::Bytes` cannot actually be indexed.
// TODO fix during the indexer refactoring described in #276

**Comment by fulmicoton** [comment]
Please edit `CHANGELOG.md` for `0.6.0` and make sure to add your username on the feature you add. It makes the release process much smoother.

**Comment by jason-wolfe** [comment]
Ready for merge from my side, pending CR.

File: src/postings/postings_writer.rs (hunk 7 above)
Comment (jason-wolfe) [reply src/postings/postings_writer.rs]: Done

**Comment by fulmicoton** [comment]
Sweet． Will have at that tomorrow

File: src/indexer/merger.rs [hunk 9]
```diff
@@ -284,25 +287,27 @@ impl IndexMerger {
...
         // what should be the bit length use for bitpacking.
         for reader in &self.readers {
-            let multi_ff_reader = reader.multi_fast_field_reader::<u64>(field)?;
+            let idx_reader = reader.fast_field_reader_with_idx::<u64>(field, 0)?;
             if let Some(delete_bitset) = reader.delete_bitset() {
                 for doc in 0u32..reader.max_doc() {
                     if !delete_bitset.is_deleted(doc) {
-                        total_num_vals += multi_ff_reader.num_vals(doc) as u64;
+                        let start = idx_reader.get(doc);
+                        let end = idx_reader.get(doc + 1);
```
Comment (fulmicoton) [inline src/indexer/merger.rs:305]: why did you remove `.num_vals` ?

File: src/indexer/merger.rs [hunk 10]
```diff
@@ -311,11 +316,13 @@ impl IndexMerger {
...
         let mut idx = 0;
         for reader in &self.readers {
-            let multi_ff_reader = reader.multi_fast_field_reader::<u64>(field)?;
+            let idx_reader = reader.fast_field_reader_with_idx::<u64>(field, 0)?;
             for doc in 0u32..reader.max_doc() {
                 if !reader.is_deleted(doc) {
                     serialize_idx.add_val(idx)?;
-                    idx += multi_ff_reader.num_vals(doc) as u64;
+                    let start = idx_reader.get(doc);
+                    let end = idx_reader.get(doc + 1);
```
Comment (fulmicoton) [inline src/indexer/merger.rs:324]: why did you remove .num_vals ?

File: src/core/segment_reader.rs [hunk 11]
```diff
@@ -149,6 +150,21 @@ impl SegmentReader {
         }
     }
 
+    /// Accessor to the `BytesFastFieldReader` associated to a given `Field`.
+    /// May panic if the field is not a Bytes fast field.
```
Comment (fulmicoton) [inline src/core/segment_reader.rs]: Does it panic though?

File: src/core/segment_reader.rs [hunk 12]
```diff
@@ -149,6 +150,21 @@ impl SegmentReader {
...
+    pub fn bytes_fast_field_reader(&self, field: Field) -> fastfield::Result<BytesFastFieldReader> {
+        let field_entry = self.schema.get_field_entry(field);
```
Comment (fulmicoton) [inline src/core/segment_reader.rs]: maybe test if the type is right now that you have the field entry, and return the error here.

The benefit is then that you `panic!` only if the index is corrupted, but you return a nice error if the field just does not have the right type.

File: src/indexer/merger.rs (hunk 9 above)
Comment (jason-wolfe) [reply src/indexer/merger.rs:305]: Because it's no longer used. I could move it to FastFieldReader<u64> (which is all this method uses now), but it's no longer used by anything on the multi reader. It seems weird on FastFieldReader<u64>, however, because it's only meaningful for indexes.

File: src/core/segment_reader.rs [hunk 13]
```diff
@@ -149,6 +150,21 @@ impl SegmentReader {
...
+        let idx_reader = self.fast_fields_composite
+            .open_read_with_idx(field, 0)
+            .ok_or_else(|| FastFieldNotAvailableError::new(field_entry))
+            .map(FastFieldReader::open)?;
+        let values = self.fast_fields_composite
+            .open_read_with_idx(field, 1)
+            .ok_or_else(|| FastFieldNotAvailableError::new(field_entry))
+            .map(|x| OwningRef::new(x).map(|source| &source[..]))?;
```
[the remark this answers is not in the document]
Comment (jason-wolfe) [reply src/core/segment_reader.rs]: Sure.

File: src/core/segment_reader.rs (hunk 11 above)
Comment (jason-wolfe) [reply src/core/segment_reader.rs]: It doesn't. Will remove.

File: src/indexer/merger.rs (hunk 10 above)
Comment (jason-wolfe) [reply src/indexer/merger.rs:324]: Answered in other comment.

File: src/indexer/merger.rs (hunk 9 above)
Comment (fulmicoton) [reply src/indexer/merger.rs:305]: ok I see

File: src/core/segment_reader.rs (hunk 11 above)
Comment (fulmicoton) [reply src/core/segment_reader.rs]: Right now I think it does panic but in some specific corner case.. You can remove the panic if you address the other comment below.

File: src/core/segment_reader.rs (hunk 12 above)
Comment (jason-wolfe) [reply src/core/segment_reader.rs]: Should I turn FastFieldNotAvailableError into a FastFieldError enum to indicate the fast field is the wrong type? Just saying it's not available seems confusing.
EDIT: Actually I don't think there's a good way to do this cleanly. I'll stick with FastFieldNotAvailableError. But maybe should file a follow-up to do this more nicely.

Comment (fulmicoton) [reply src/core/segment_reader.rs]: ok. I also think making the two unindistiguishable is not the best but we can keep it that way for the moment.
